omc3/kmod/analysis.py

```diff
@@ -247,49 +253,145 @@ def return_df(magnet1_df, magnet2_df, plane):
-           sign[5] * kmod_input_params.errorK * def_magnet_df.headers[K],
-           def_magnet_df.headers['LSTAR'] +
-           sign[6] * kmod_input_params.misalignment) -
-           def_magnet_df.headers[f"{AVERAGE}{BETA}{plane}"] +
-           sign[7] * foc_magnet_df.headers[f"{ERR}{AVERAGE}{BETA}{plane}"]) ** 2)
+    phase_adv_constraint = phase_constraint(kmod_input_params, plane)
+    if kmod_input_params.interaction_point:
+        BPML,BPMR = get_BPM(kmod_input_params)
+        BPM_distance = get_BPM_distance(kmod_input_params,BPML,BPMR)
+        phase_adv = phase_adv_from_kmod(BPM_distance, b, 0.0, w, 0.0)[0]
+        weight = phase_adv_constraint[2]
+    else:
+        phase_adv = 0.0
+        weight = 0
+
+    c2 = (1-weight)*(((average_beta_focussing_quadrupole(b, w, foc_magnet_df.headers['LENGTH'] +
+        sign[0] * kmod_input_params.errorL, foc_magnet_df.headers[K] +
+        sign[1] * kmod_input_params.errorK * foc_magnet_df.headers[K],
+        foc_magnet_df.headers['LSTAR'] +
+        sign[2] * kmod_input_params.misalignment) -
+        foc_magnet_df.headers[f"{AVERAGE}{BETA}{plane}"] +
+        sign[3] * foc_magnet_df.headers[f"{ERR}{AVERAGE}{BETA}{plane}"])/(def_magnet_df.headers[f"{AVERAGE}{BETA}{plane}"] + foc_magnet_df.headers[f"{AVERAGE}{BETA}{plane}"])/2.0) ** 2 +
+        ((average_beta_defocussing_quadrupole(b, -w, def_magnet_df.headers['LENGTH'] +
+        sign[4] * kmod_input_params.errorL, def_magnet_df.headers[K] +
+        sign[5] * kmod_input_params.errorK * def_magnet_df.headers[K],
+        def_magnet_df.headers['LSTAR'] +
+        sign[6] * kmod_input_params.misalignment) -
+        def_magnet_df.headers[f"{AVERAGE}{BETA}{plane}"] +
+        sign[7] * def_magnet_df.headers[f"{ERR}{AVERAGE}{BETA}{plane}"])/(foc_magnet_df.headers[f"{AVERAGE}{BETA}{plane}"] + def_magnet_df.headers[f"{AVERAGE}{BETA}{plane}"])/2.0) ** 2) + \
+        weight*(((phase_adv - (phase_adv_constraint[0]+sign[8]*phase_adv_constraint[1]))/phase_adv_constraint[0])**2)
 
     return c2
 
+def get_beta_waist(magnet1_df, magnet2_df, kmod_input_params, plane, betastar_required):
 
-def get_beta_waist(magnet1_df, magnet2_df, kmod_input_params, plane):
-    n = 8
+    n = 9
     sign = return_sign_for_err(n)
     foc_magnet_df, def_magnet_df = return_df(magnet1_df, magnet2_df, plane)
     results = np.zeros((2*n+1, 2))
+
     for i, s in enumerate(sign):
-        def fun(x): return chi2(x, foc_magnet_df, def_magnet_df, plane, kmod_input_params, s)
+
+        def fun(x): return chi2(x, foc_magnet_df, def_magnet_df, plane, kmod_input_params, s, betastar_required)
         fitresults = scipy.optimize.minimize(fun=fun,
                                              x0=kmod_input_params.betastar_and_waist[plane],
                                              method='nelder-mead',
-                                             tol=1E-22)
+                                             tol=1E-9)
```

Comment:
Is there a reason to relax the tolerance?

Reply:
It was a test I forgot to put back to 1e-22